File: ViewFinderAnywhere/src/main/java/com/fezrestia/android/viewfinderanywhere/receiver/OverlayViewFinderTriggerReceiver.java
package com.fezrestia.android.viewfinderanywhere.receiver;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;

import com.fezrestia.android.lib.util.log.Log;
import com.fezrestia.android.viewfinderanywhere.ViewFinderAnywhereConstants;
import com.fezrestia.android.viewfinderanywhere.control.OnOffTrigger;

public class OverlayViewFinderTriggerReceiver extends BroadcastReceiver {
    // Log tag.
    private static final String TAG = "OverlayViewFinderTriggerReceiver";

    /**
     * Register this receiver to system.
     *
     * @param context Master context
     */
    public void register(Context context) {
        IntentFilter filter = new IntentFilter();
        filter.addAction(ViewFinderAnywhereConstants.INTENT_ACTION_TOGGLE_OVERLAY_VISIBILITY);
        filter.addAction(Intent.ACTION_SCREEN_ON);
        filter.addAction(Intent.ACTION_SCREEN_OFF);

        context.registerReceiver(this, filter);
    }

    /**
     * Unregister this receiver from system.
     *
     * @param context Master context
     */
    public void unregister(Context context) {
        context.unregisterReceiver(this);
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        if (Log.IS_DEBUG) Log.logDebug(TAG, "onReceive()");

        String action = intent.getAction();
        if (Log.IS_DEBUG) Log.logDebug(TAG, "ACTION = " + action);

        if (action == null) {
            if (Log.IS_DEBUG) Log.logDebug(TAG, "ACTION is NULL.");
            // NOP.
        } else switch (action) {
            case ViewFinderAnywhereConstants.INTENT_ACTION_TOGGLE_OVERLAY_VISIBILITY: {
                OnOffTrigger.requestToggleVisibility(context);
            }
            break;

            case Intent.ACTION_SCREEN_ON: {
                OnOffTrigger.onScreenOn(context);
            }
            break;

            case Intent.ACTION_SCREEN_OFF: {
                OnOffTrigger.onScreenOff(context);
            }
            break;

            default:
                // Unexpected Action.
                throw new IllegalArgumentException("Unexpected Action : " + action);
        }
    }
}

File: ViewFinderAnywhere/src/main/java/com/fezrestia/android/viewfinderanywhere/receiver/OverlayViewFinderTriggerReceiver.kt
@file:Suppress("PrivatePropertyName", "ConstantConditionIf")

package com.fezrestia.android.viewfinderanywhere.receiver

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Build

import com.fezrestia.android.lib.util.log.IS_DEBUG
import com.fezrestia.android.lib.util.log.logD
import com.fezrestia.android.viewfinderanywhere.App
import com.fezrestia.android.viewfinderanywhere.Constants
import com.fezrestia.android.viewfinderanywhere.control.OnOffTrigger

class OverlayViewFinderTriggerReceiver : BroadcastReceiver() {
    /**
     * Register this receiver to system.
     *
     * @param context Master context
     */
    fun register(context: Context) {
        val filter = IntentFilter()
        filter.addAction(Constants.INTENT_ACTION_TOGGLE_OVERLAY_VISIBILITY)
        filter.addAction(Constants.INTENT_ACTION_TOGGLE_OVERLAY_ENABLE_DISABLE)
        filter.addAction(Intent.ACTION_SCREEN_ON)
        filter.addAction(Intent.ACTION_SCREEN_OFF)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            context.registerReceiver(this, filter, Context.RECEIVER_NOT_EXPORTED)
        } else {
            // noinspection UnspecifiedRegisterReceiverFlag
            context.registerReceiver(this, filter)
        }
    }

    /**
     * Unregister this receiver from system.
     *
     * @param context Master context
     */
    fun unregister(context: Context) {
        context.unregisterReceiver(this)
    }

    override fun onReceive(context: Context, intent: Intent) {
        if (IS_DEBUG) logD(TAG, "onReceive()")
        if (IS_DEBUG) logD(TAG, "ACTION = ${intent.action}")

        when (intent.action) {
            Constants.INTENT_ACTION_TOGGLE_OVERLAY_VISIBILITY -> {
                OnOffTrigger.requestToggleVisibility(context)
            }

            Constants.INTENT_ACTION_TOGGLE_OVERLAY_ENABLE_DISABLE -> {
                if (App.isOverlayServiceActive) {
                    OnOffTrigger.requestStop(context)
                } else {
                    OnOffTrigger.requestStart(context)
                }
            }

            Intent.ACTION_SCREEN_ON -> {
                OnOffTrigger.onScreenOn(context)
            }

            Intent.ACTION_SCREEN_OFF -> {
                OnOffTrigger.onScreenOff(context)
            }

            else -> {
                // Unexpected Action.
                throw IllegalArgumentException("Unexpected Action = ${intent.action}")
            }
        }
    }

    companion object {
        private const val TAG = "OverlayViewFinderTriggerReceiver"
    }
}
